## Hard Drive Data

This is an early exploration on the hard drive 
dataset 



In [4]:
## Import the libraries 
import pandas as pd 
import numpy as np 
from matplotlib import pyplot as plt 

%matplotlib inline 

In [3]:
data = pd.read_csv("data/harddrive.csv")

In [5]:
data.head(10)

date    serial_number                    model  capacity_bytes  \
0  2016-01-01   MJ0351YNG9Z0XA  Hitachi HDS5C3030ALA630   1.482490e-311   
1  2016-01-01         Z305B2QN              ST4000DM000   1.976651e-311   
2  2016-01-01   MJ0351YNG9Z7LA  Hitachi HDS5C3030ALA630   1.482490e-311   
3  2016-01-01   MJ0351YNGABYAA  Hitachi HDS5C3030ALA630   1.482490e-311   
4  2016-01-01  WD-WMC4N2899475             WDC WD30EFRX   1.482490e-311   
5  2016-01-01   PL1321LAG34XWH  Hitachi HDS5C4040ALE630   1.976651e-311   
6  2016-01-01   MJ1323YNG1TJ7C  Hitachi HDS5C3030ALA630   1.482490e-311   
7  2016-01-01         S300Z6K9              ST4000DM000   1.976651e-311   
8  2016-01-01         S300Z6K8              ST4000DM000   1.976651e-311   
9  2016-01-01         Z302PGH8              ST4000DM000   1.976651e-311   

   failure  smart_1_normalized  smart_1_raw  smart_2_normalized  smart_2_raw  \
0        0                 100            0               135.0        108.0   
1        0                 113     54551400                 NaN          NaN   
2        0                 100            0               136.0        104.0   
3        0                 100            0               136.0        104.0   
4        0                 200            0                 NaN          NaN   
5        0                 100            0               134.0        101.0   
6        0                 100            0               100.0          0.0   
7        0                 107     13985080                 NaN          NaN   
8        0                 116    109242152                 NaN          NaN   
9        0                 112     46112000                 NaN          NaN   

   smart_3_normalized      ...        smart_250_normalized  smart_250_raw  \
0                 143      ...                         NaN            NaN   
1                  96      ...                         NaN            NaN   
2                 124      ...                         NaN            NaN   
3                 137      ...                         NaN            NaN   
4                 175      ...                         NaN            NaN   
5                 131      ...                         NaN            NaN   
6                 122      ...                         NaN            NaN   
7                  97      ...                         NaN            NaN   
8                  97      ...                         NaN            NaN   
9                  92      ...                         NaN            NaN   

   smart_251_normalized  smart_251_raw  smart_252_normalized  smart_252_raw  \
0                   NaN            NaN                   NaN            NaN   
1                   NaN            NaN                   NaN            NaN   
2                   NaN            NaN                   NaN            NaN   
3                   NaN            NaN                   NaN            NaN   
4                   NaN            NaN                   NaN            NaN   
5                   NaN            NaN                   NaN            NaN   
6                   NaN            NaN                   NaN            NaN   
7                   NaN            NaN                   NaN            NaN   
8                   NaN            NaN                   NaN            NaN   
9                   NaN            NaN                   NaN            NaN   

   smart_254_normalized  smart_254_raw  smart_255_normalized  smart_255_raw  
0                   NaN            NaN                   NaN            NaN  
1                   NaN            NaN                   NaN            NaN  
2                   NaN            NaN                   NaN            NaN  
3                   NaN            NaN                   NaN            NaN  
4                   NaN            NaN                   NaN            NaN  
5                   NaN            NaN                   NaN            NaN  
6                   NaN            NaN

In [7]:
## The number of failures
data.failure.value_counts()

0    3179080
1        215
Name: failure, dtype: int64

There are too few failures - only 215 overall. 
Lets try and see how many failures there are per model 

In [8]:
models = data.model.unique()
for m in models: 
    try: 
        print("The model, ", m, " has ", data[data["model"]==m].failure.value_counts()[1], " failures")
    except KeyError:
        print("The model, ", m, " has no failures")

The model,  Hitachi HDS5C3030ALA630  has  5  failures
The model,  ST4000DM000  has  139  failures
The model,  WDC WD30EFRX  has  6  failures
The model,  Hitachi HDS5C4040ALE630  has  4  failures
The model,  HGST HMS5C4040ALE640  has  2  failures
The model,  HGST HMS5C4040BLE640  has  1  failures
The model,  ST6000DX000  has  2  failures
The model,  Hitachi HDS723030ALA640  has no failures
The model,  Hitachi HDS722020ALA330  has  13  failures
The model,  ST31500541AS  has no failures
The model,  ST4000DX000  has  2  failures
The model,  WDC WD60EFRX  has  1  failures
The model,  WDC WD5000LPVX  has no failures
The model,  HGST HUH728080ALE600  has no failures
The model,  WDC WD1600AAJS  has  2  failures
The model,  ST320LT007  has  15  failures
The model,  ST500LM012 HN  has no failures
The model,  TOSHIBA MD04ABA500V  has  1  failures
The model,  ST9320325AS  has no failures
The model,  TOSHIBA MD04ABA400V  has no failures
The model,  TOSHIBA DT01ACA300  has  1  failures
The model,  H

The model ST4000DM000 has the most number of failures. Lets have a look at it closer 

In [12]:
## Look at all the features
data[data["model"]=="ST4000DM000"].columns

Index(['date', 'serial_number', 'model', 'capacity_bytes', 'failure',
       'smart_1_normalized', 'smart_1_raw', 'smart_2_normalized',
       'smart_2_raw', 'smart_3_normalized', 'smart_3_raw',
       'smart_4_normalized', 'smart_4_raw', 'smart_5_normalized',
       'smart_5_raw', 'smart_7_normalized', 'smart_7_raw',
       'smart_8_normalized', 'smart_8_raw', 'smart_9_normalized',
       'smart_9_raw', 'smart_10_normalized', 'smart_10_raw',
       'smart_11_normalized', 'smart_11_raw', 'smart_12_normalized',
       'smart_12_raw', 'smart_13_normalized', 'smart_13_raw',
       'smart_15_normalized', 'smart_15_raw', 'smart_22_normalized',
       'smart_22_raw', 'smart_183_normalized', 'smart_183_raw',
       'smart_184_normalized', 'smart_184_raw', 'smart_187_normalized',
       'smart_187_raw', 'smart_188_normalized', 'smart_188_raw',
       'smart_189_normalized', 'smart_189_raw', 'smart_190_normalized',
       'smart_190_raw', 'smart_191_normalized', 'smart_191_raw',
       'smart_1

In [78]:
## Lets look at the SMART features - only the raw ones
smart_features_raw = [x for x in data.columns if "smart" in x and "raw" in x]
print("Number of raw smart features: ", len(smart_features_raw))
smart_features_norm = [x for x in data.columns if "smart" in x and "normalized" in x]
print("Number of normed smart features: ", len(smart_features_norm))


Number of raw smart features:  45
Number of normed smart features:  45


In [73]:
type(data.smart_199_raw)
len(data.smart_199_raw), data.shape[0]
print(data.smart_199_raw.shape[0])
data.smart_199_raw.name
data.smart_255_raw.isnull().sum()

3179295


3179295

In [56]:
def howmanynulls(a, df=None):
    if type(a) != list: 
        print("% of null in ", a.name, " is ", a.isnull().sum()/a.shape[0] * 100) 
    elif type(a) == list:
        for a1 in a:
            howmanynulls(df[a1])
    

In [74]:
howmanynulls(data.smart_255_raw)
howmanynulls(smart_features_raw, df=data)

% of null in  smart_255_raw  is  100.0
% of null in  smart_1_raw  is  0.0
% of null in  smart_2_raw  is  60.3963142772
% of null in  smart_3_raw  is  0.0
% of null in  smart_4_raw  is  0.0
% of null in  smart_5_raw  is  0.0
% of null in  smart_7_raw  is  0.0
% of null in  smart_8_raw  is  60.3961255561
% of null in  smart_9_raw  is  0.0
% of null in  smart_10_raw  is  0.0
% of null in  smart_11_raw  is  95.382183786
% of null in  smart_12_raw  is  0.0
% of null in  smart_13_raw  is  99.9998112789
% of null in  smart_15_raw  is  100.0
% of null in  smart_22_raw  is  99.9263987771
% of null in  smart_183_raw  is  43.4007539407
% of null in  smart_184_raw  is  43.1787235849
% of null in  smart_187_raw  is  43.1787235849
% of null in  smart_188_raw  is  0.0
% of null in  smart_189_raw  is  43.1805478888
% of null in  smart_190_raw  is  43.1780316076
% of null in  smart_191_raw  is  41.843899355
% of null in  smart_192_raw  is  0.301356118259
% of null in  smart_193_raw  is  1.34231016625
%

In [75]:
howmanynulls(smart_features_norm, df=data)

% of null in  smart_1_normalized  is  0.0
% of null in  smart_2_normalized  is  60.3963142772
% of null in  smart_3_normalized  is  0.0
% of null in  smart_4_normalized  is  0.0
% of null in  smart_5_normalized  is  0.0
% of null in  smart_7_normalized  is  0.0
% of null in  smart_8_normalized  is  60.3961255561
% of null in  smart_9_normalized  is  0.0
% of null in  smart_10_normalized  is  0.0
% of null in  smart_11_normalized  is  95.382183786
% of null in  smart_12_normalized  is  0.0
% of null in  smart_13_normalized  is  99.9998112789
% of null in  smart_15_normalized  is  100.0
% of null in  smart_22_normalized  is  99.9263987771
% of null in  smart_183_normalized  is  43.4007539407
% of null in  smart_184_normalized  is  43.1787235849
% of null in  smart_187_normalized  is  43.1787235849
% of null in  smart_188_normalized  is  43.1787235849
% of null in  smart_189_normalized  is  43.1805478888
% of null in  smart_190_normalized  is  43.1780316076
% of null in  smart_191_normali

So it turns out some of the features are just null - no matter what type of hard drive it is. 

In which case, maybe we can get rid of those.
So the ones to get rid of are: 
200, 201, 220, 222, 223, 224,225, 226, 250, 251,252, 254, 255, 
195, 22, 11



In [80]:
for i in [200, 201, 220, 222, 223, 224,225, 226, 250, 251,252, 254, 255, 195, 22, 11]:
    try:
        smart_features_raw.remove('smart_'+str(i)+'_raw')
    except ValueError: 
        print("Index not found: ", i)
    


In [82]:
smart_features_raw

['smart_1_raw',
 'smart_2_raw',
 'smart_3_raw',
 'smart_4_raw',
 'smart_5_raw',
 'smart_7_raw',
 'smart_8_raw',
 'smart_9_raw',
 'smart_10_raw',
 'smart_12_raw',
 'smart_13_raw',
 'smart_15_raw',
 'smart_183_raw',
 'smart_184_raw',
 'smart_187_raw',
 'smart_188_raw',
 'smart_189_raw',
 'smart_190_raw',
 'smart_191_raw',
 'smart_192_raw',
 'smart_193_raw',
 'smart_194_raw',
 'smart_196_raw',
 'smart_197_raw',
 'smart_198_raw',
 'smart_199_raw',
 'smart_240_raw',
 'smart_241_raw',
 'smart_242_raw']

In [2]:
def mmm(a, df=None):
    if type(a) != list:
        print("For ", a.name, "mean: ", str(a.mean()), " median: ", str(a.median()), " mode: ", str(a.mode()))
    else:
        for a1 in a:
            mmm(df[a1])


In [3]:
mmm(smart_features_raw, df=data) 

NameError: name 'smart_features_raw' is not defined